python/ray/experimental/workflow/tests/test_basic_workflows.py

```diff
@@ -171,3 +172,64 @@ def test_run_or_resume_during_running(ray_start_regular_shared):
     with pytest.raises(ValueError):
         workflow.resume(workflow_id="running_workflow")
     assert ray.get(output) == "[source1][append1][append2]"
+
+
+@pytest.mark.parametrize(
+    "ray_start_regular_shared", [{
+        "namespace": "workflow"
+    }], indirect=True)
+def test_step_failure(ray_start_regular_shared, tmp_path):
+    (tmp_path / "test").write_text("0")
+
+    @workflow.step
+    def unstable_step():
+        v = int((tmp_path / "test").read_text())
+        (tmp_path / "test").write_text(f"{v + 1}")
+        if v < 10:
+            raise ValueError("Invalid")
+        return v
+
+    with pytest.raises(Exception):
+        unstable_step.options(step_max_retries=-1).step().run()
```

Comment:
How about just max_retries?

Reply:
I use step_max_retries to differentiate with ray internal errors on purpose. I gave some thought to this and I think it's good not to mix them.

Usually, when users specify the retry number here, they actually want to retry the application logic. Right now, max_retries has been used by ray, so I decide to use step_max_retries.

I'm thinking about hiding the ray internal retries from the user, but I'm not sure whether it's a good direction or not.

Right now, we don't have an easy way to detect ray internal errors without `ray.get` which I think is something missing and we probably want to add that later. It's an important feature that allows the workflow to detect errors in ray in a better way.